### PR TITLE
change the representation of UID

### PR DESCRIPTION
**Pull request by InnocentBug**
# Description
This change the JSON representatin of UID node.

## Changes

From `{"node": ["NodeType"], "uid": "_:asdfasdfasdf"}` to  `{"uid": "_:asdfasdfasdf"}`.
## Tests

Some are passing, some might be failing.
Before we go through the effort and make all the tests pass, we should find out which version of UID is correct.

## Known Issues

Some tests might fail.

## Checklist

- [x] My name is on the list of contributors (`CONTRIBUTORS.md`) in the pull request source branch.
- [x] I have updated the documentation to reflect my changes.

### CODE REVIEW

**Comment by InnocentBug** [comment]
Ok, according to brili, this is the correct way now of representing UID in JSON.

**Comment by InnocentBug** [comment]
This should pass the automated tests here.

However, locally I am hitting integration errors that weren't there before.

Not sure if they from this change or a change in the back end.

File: tests/conftest.py [hunk 1]
```diff
@@ -17,9 +17,21 @@
...
+
+def _get_cript_tests_env() -> bool:
+    """
+    A little helper function, that ensure to mostly default to True.
+    Except if an explicit spelling of False is detected.
+    """
+    env_var = os.getenv("CRIPT_TESTS")
+    if env_var in ("false", "FALSE", "False"):
+        return False
+    return True
```
Comment (nh916) [inline tests/conftest.py:29]:
```python
def _get_cript_tests_env() -> bool:
    """
    A little helper function, that ensure to mostly default to True.
    Except if an explicit spelling of False is detected.
    """
    # get env var string, convert to title case, compare that it is equal to the string "True"
    return os.getenv("CRIPT_TESTS").title().strip() == "True"
```

the current version would return True if `CRIPT_TESTS` has `FalSE`

File: tests/conftest.py [hunk 2]
```diff
@@ -17,9 +17,21 @@
...
+
+
 # flip integration tests ON or OFF with this boolean
 # automatically gets value env vars to run integration tests
-HAS_INTEGRATION_TESTS_ENABLED: bool = os.getenv("CRIPT_TESTS").title() == "True"
+HAS_INTEGRATION_TESTS_ENABLED: bool = _get_cript_tests_env()
```
Comment (nh916) [inline tests/conftest.py:34]: why was this change necessary? the previous code functioned correctly and was clean

**Comment by nh916** [comment]
> This should pass the automated tests here.
> 
> However, locally I am hitting integration errors that weren't there before.
> 
> Not sure if they from this change or a change in the back end.

as far as I know Brili said that the development backend is not changing and is not being maintained any more, so I would guess that it is from this change

File: tests/conftest.py (hunk 1 above)
Comment (InnocentBug) [reply tests/conftest.py:29]: Correct, I don't think that is very likely though, that people spell it that way.

File: tests/conftest.py (hunk 2 above)
Comment (InnocentBug) [reply tests/conftest.py:34]: It was necessary, because the old version didn't tolerate no specification of the ENV variable.

That case `os.getenv("CRIPT_TESTS")` returns `None`.
And you can't call `.title()` on a `None` object.

Comment (nh916) [reply tests/conftest.py:34]: can't you just add a small try except block then?

Comment (InnocentBug) [reply tests/conftest.py:34]: an except block would work too, but I don't like the logic of an except block here.

I don't really care though.
This is a tiny detail.

**Comment by nh916** [comment]
> "Before we go through the effort and make all the tests pass, we should find out which version of UID is correct."

I think Brili said it here: https://mit-olsenlab.slack.com/archives/C03T4S02HT4/p1691013559647979?thread_ts=1691013117.144499&cid=C03T4S02HT4

do we still have more questions?

**Comment by nh916** [comment]
I can try to take a look at the errors and see what is going on

**Comment by InnocentBug** [comment]
> > "Before we go through the effort and make all the tests pass, we should find out which version of UID is correct."
> 
> I think Brili said it here: https://mit-olsenlab.slack.com/archives/C03T4S02HT4/p1691013559647979?thread_ts=1691013117.144499&cid=C03T4S02HT4
> 
> do we still have more questions?

I can't open this link. But yes Brili confirmed it now.
We should merge this.

**Comment by InnocentBug** [comment]
> I can try to take a look at the errors and see what is going on

Thank you

**Comment by InnocentBug** [comment]
> I can try to take a look at the errors and see what is going on

It is for example in `tests/nodes/primary_nodes/test_computational_process.py::test_integration_computational_process`


And it doesn't make sense to me.
Because, we are sending a POST and locally we are validating the JSON against POST and it passes.
But the response says, that it violates the schema:

```
cript.api.exceptions.CRIPTAPISaveError: API responded to POST with 'http:400 {'node': ['Experiment'], 'uid': '_:8bf04055-931c-4806-9f44-002e3cce6aa1', 'uuid': '8bf04055-931c-4806-9f44-002e3cce6aa1', 'name': 'my experiment name', 'computation_process': [{'node': ['ComputationProcess'], 'uid': '_:f307d0ea-bb52-4d32-84a3-5d86717af88f', 'uuid': 'f307d0ea-bb52-4d32-84a3-5d86717af88f', 'name': 'my computational process node name', 'type': 'cross_linking', 'input_data': [{'node': ['Data'], 'uid': '_:e5d2144c-6251-469f-ad62-786c1add754a', 'uuid': 'e5d2144c-6251-469f-ad62-786c1add754a', 'name': 'my data name', 'type': 'afm_amp', 'file': [{'node': ['File'], 'uid': '_:6f17d3b3-7df3-4b60-a1a9-2548f492e54b', 'uuid': '6f17d3b3-7df3-4b60-a1a9-2548f492e54b', 'name': 'my complex file node fixture', 'source': 'https://criptapp.org', 'type': 'calibration', 'extension': '.csv', 'data_dictionary': "my file's data dictionary", 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'ingredient': [{'node': ['Ingredient'], 'uid': '_:ebd3f292-2334-4e0f-923e-4b6404854e39', 'uuid': 'ebd3f292-2334-4e0f-923e-4b6404854e39', 'material': {'uuid': '73464d42-ca67-4ce2-b889-983804afdec9'}, 'quantity': [{'node': ['Quantity'], 'uid': '_:6bd99252-0d77-47c3-b991-efa649bfdaad', 'uuid': '6bd99252-0d77-47c3-b991-efa649bfdaad', 'key': 'mass', 'value': 11.2, 'unit': 'kg', 'uncertainty': 0.2, 'uncertainty_type': 'stdev', 'model_version': '1.0.0', 'updated_at': 'now()', 'created_at': 'now()'}], 'keyword': ['catalyst'], 'model_version': '1.0.0', 'updated_at': 'now()', 'created_at': 'now()'}], 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'data': [{'uid': '_:e5d2144c-6251-469f-ad62-786c1add754a', 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'uuid': 'uuid()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'} is not valid under any of the given schemas at path: properties/collection/items/properties/experiment/items' data: {"node": ["Project"], "uid": "_:7a832c25-13d2-49ac-808d-f1387ee6a321", "uuid": "7a832c25-13d2-49ac-808d-f1387ee6a321", "name": "test_integration_computation_process_name_97d1a85c8e7647589d75083a18714a91", "collection": [{"node": ["Collection"], "uid": "_:f93117a9-53c7-4cc6-8b3c-b1920051be70", "uuid": "f93117a9-53c7-4cc6-8b3c-b1920051be70", "name": "my collection name", "experiment": [{"node": ["Experiment"], "uid": "_:8bf04055-931c-4806-9f44-002e3cce6aa1", "uuid": "8bf04055-931c-4806-9f44-002e3cce6aa1", "name": "my experiment name", "computation_process": [{"node": ["ComputationProcess"], "uid": "_:f307d0ea-bb52-4d32-84a3-5d86717af88f", "uuid": "f307d0ea-bb52-4d32-84a3-5d86717af88f", "name": "my computational process node name", "type": "cross_linking", "input_data": [{"node": ["Data"], "uid": "_:e5d2144c-6251-469f-ad62-786c1add754a", "uuid": "e5d2144c-6251-469f-ad62-786c1add754a", "name": "my data name", "type": "afm_amp", "file": [{"node": ["File"], "uid": "_:6f17d3b3-7df3-4b60-a1a9-2548f492e54b", "uuid": "6f17d3b3-7df3-4b60-a1a9-2548f492e54b", "name": "my complex file node fixture", "source": "https://criptapp.org", "type": "calibration", "extension": ".csv", "data_dictionary": "my file's data dictionary"}]}], "ingredient": [{"node": ["Ingredient"], "uid": "_:ebd3f292-2334-4e0f-923e-4b6404854e39", "uuid": "ebd3f292-2334-4e0f-923e-4b6404854e39", "material": {"uuid": "73464d42-ca67-4ce2-b889-983804afdec9"}, "quantity": [{"node": ["Quantity"], "uid": "_:6bd99252-0d77-47c3-b991-efa649bfdaad", "uuid": "6bd99252-0d77-47c3-b991-efa649bfdaad", "key": "mass", "value": 11.2, "unit": "kg", "uncertainty": 0.2, "uncertainty_type": "stdev"}], "keyword": ["catalyst"]}]}], "data": [{"uid": "_:e5d2144c-6251-469f-ad62-786c1add754a"}]}], "inventory": [], "citation": []}], "material": [{"node": ["Material"], "uid": "_:73464d42-ca67-4ce2-b889-983804afdec9", "uuid": "73464d42-ca67-4ce2-b889-983804afdec9", "name": "my test material 5cbffc08-38b4-4bc0-9c4e-b84b643aff95_1fbb653cf7d840b6a503cb680feeebcb_220d7cb24ad948c4b51a6dd367a9ae6f", "bigsmiles": "123456"}]}
```

File: tests/conftest.py (hunk 1 above)
Comment (nh916) [reply tests/conftest.py:29]: I have a hard time accepting this, I made a PR for this issue, let me know what you think

https://github.com/C-Accel-CRIPT/Python-SDK/pull/253

**Comment by nh916** [comment]
> > I can try to take a look at the errors and see what is going on
> 
> It is for example in `tests/nodes/primary_nodes/test_computational_process.py::test_integration_computational_process`
> 
> And it doesn't make sense to me. Because, we are sending a POST and locally we are validating the JSON against POST and it passes. But the response says, that it violates the schema:
> 
> ```
> cript.api.exceptions.CRIPTAPISaveError: API responded to POST with 'http:400 {'node': ['Experiment'], 'uid': '_:8bf04055-931c-4806-9f44-002e3cce6aa1', 'uuid': '8bf04055-931c-4806-9f44-002e3cce6aa1', 'name': 'my experiment name', 'computation_process': [{'node': ['ComputationProcess'], 'uid': '_:f307d0ea-bb52-4d32-84a3-5d86717af88f', 'uuid': 'f307d0ea-bb52-4d32-84a3-5d86717af88f', 'name': 'my computational process node name', 'type': 'cross_linking', 'input_data': [{'node': ['Data'], 'uid': '_:e5d2144c-6251-469f-ad62-786c1add754a', 'uuid': 'e5d2144c-6251-469f-ad62-786c1add754a', 'name': 'my data name', 'type': 'afm_amp', 'file': [{'node': ['File'], 'uid': '_:6f17d3b3-7df3-4b60-a1a9-2548f492e54b', 'uuid': '6f17d3b3-7df3-4b60-a1a9-2548f492e54b', 'name': 'my complex file node fixture', 'source': 'https://criptapp.org', 'type': 'calibration', 'extension': '.csv', 'data_dictionary': "my file's data dictionary", 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'ingredient': [{'node': ['Ingredient'], 'uid': '_:ebd3f292-2334-4e0f-923e-4b6404854e39', 'uuid': 'ebd3f292-2334-4e0f-923e-4b6404854e39', 'material': {'uuid': '73464d42-ca67-4ce2-b889-983804afdec9'}, 'quantity': [{'node': ['Quantity'], 'uid': '_:6bd99252-0d77-47c3-b991-efa649bfdaad', 'uuid': '6bd99252-0d77-47c3-b991-efa649bfdaad', 'key': 'mass', 'value': 11.2, 'unit': 'kg', 'uncertainty': 0.2, 'uncertainty_type': 'stdev', 'model_version': '1.0.0', 'updated_at': 'now()', 'created_at': 'now()'}], 'keyword': ['catalyst'], 'model_version': '1.0.0', 'updated_at': 'now()', 'created_at': 'now()'}], 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'data': [{'uid': '_:e5d2144c-6251-469f-ad62-786c1add754a', 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'uuid': 'uuid()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'}], 'model_version': '1.0.0', 'updated_by': {'uuid': 'user_uuid()'}, 'updated_at': 'now()', 'public': False, 'locked': False, 'notes': '', 'created_by': {'uuid': 'user_uuid()'}, 'created_at': 'now()'} is not valid under any of the given schemas at path: properties/collection/items/properties/experiment/items' data: {"node": ["Project"], "uid": "_:7a832c25-13d2-49ac-808d-f1387ee6a321", "uuid": "7a832c25-13d2-49ac-808d-f1387ee6a321", "name": "test_integration_computation_process_name_97d1a85c8e7647589d75083a18714a91", "collection": [{"node": ["Collection"], "uid": "_:f93117a9-53c7-4cc6-8b3c-b1920051be70", "uuid": "f93117a9-53c7-4cc6-8b3c-b1920051be70", "name": "my collection name", "experiment": [{"node": ["Experiment"], "uid": "_:8bf04055-931c-4806-9f44-002e3cce6aa1", "uuid": "8bf04055-931c-4806-9f44-002e3cce6aa1", "name": "my experiment name", "computation_process": [{"node": ["ComputationProcess"], "uid": "_:f307d0ea-bb52-4d32-84a3-5d86717af88f", "uuid": "f307d0ea-bb52-4d32-84a3-5d86717af88f", "name": "my computational process node name", "type": "cross_linking", "input_data": [{"node": ["Data"], "uid": "_:e5d2144c-6251-469f-ad62-786c1add754a", "uuid": "e5d2144c-6251-469f-ad62-786c1add754a", "name": "my data name", "type": "afm_amp", "file": [{"node": ["File"], "uid": "_:6f17d3b3-7df3-4b60-a1a9-2548f492e54b", "uuid": "6f17d3b3-7df3-4b60-a1a9-2548f492e54b", "name": "my complex file node fixture", "source": "https://criptapp.org", "type": "calibration", "extension": ".csv", "data_dictionary": "my file's data dictionary"}]}], "ingredient": [{"node": ["Ingredient"], "uid": "_:ebd3f292-2334-4e0f-923e-4b6404854e39", "uuid": "ebd3f292-2334-4e0f-923e-4b6404854e39", "material": {"uuid": "73464d42-ca67-4ce2-b889-983804afdec9"}, "quantity": [{"node": ["Quantity"], "uid": "_:6bd99252-0d77-47c3-b991-efa649bfdaad", "uuid": "6bd99252-0d77-47c3-b991-efa649bfdaad", "key": "mass", "value": 11.2, "unit": "kg", "uncertainty": 0.2, "uncertainty_type": "stdev"}], "keyword": ["catalyst"]}]}], "data": [{"uid": "_:e5d2144c-6251-469f-ad62-786c1add754a"}]}], "inventory": [], "citation": []}], "material": [{"node": ["Material"], "uid": "_:73464d42-ca67-4ce2-b889-983804afdec9", "uuid": "73464d42-ca67-4ce2-b889-983804afdec9", "name": "my test material 5cbffc08-38b4-4bc0-9c4e-b84b643aff95_1fbb653cf7d840b6a503cb680feeebcb_220d7cb24ad948c4b51a6dd367a9ae6f", "bigsmiles": "123456"}]}
> ```

that error is very hard to read for me and I'm going to guess it will be very hard for the user to read and understand as well. I would very strongly recommend that we clean up this error in the future to something that is more helpful and concise. Currently, I have to figure out how to format and read this error message before I can even figure out what went wrong and how to debug it.

From quickly looking at the API response, it reminds me of the old API error that we encountered where it would simply parse the UUID incorrectly and throw an error, it wouldn't shock me if this was the same issue.

I will look into it more in the morning as it is currently 9 PM PST. 

> Note: giant JSON errors makes it harder to debug an issue and small JSON and trials makes it easier to find errors and debug them. When there is too many errors and things going on it is harder to isolate the problem.

**Comment by nh916** [comment]
I think I found the bug. Ingredient node is serializing to a full material node where it should serialize to a UUID

## Wrong
```json
{
   "node":[
      "Ingredient"
   ],
   "uid":"_:3b45e30f-f15e-4290-8c58-048b846566f9",
   "uuid":"3b45e30f-f15e-4290-8c58-048b846566f9",
   "material":{
      "node":[
         "Material"
      ],
      "uid":"_:676e3cde-54fa-4dc5-848a-bbd9e229108c",
      "uuid":"676e3cde-54fa-4dc5-848a-bbd9e229108c",
      "name":"my test material 0b574777-f154-4e31-9188-8e4f3e516a42_5465890cde6740f1afb4f1b184a2bd07_6814b50b31da4aa9940c44fa3e714415",
      "bigsmiles":"123456"
   },
   "quantity":[
      {
         "node":[
            "Quantity"
         ],
         "uid":"_:b86fb4c3-2a6e-40c0-a0f3-1260e550a51f",
         "uuid":"b86fb4c3-2a6e-40c0-a0f3-1260e550a51f",
         "key":"mass",
         "value":11.2,
         "unit":"kg",
         "uncertainty":0.2,
         "uncertainty_type":"stdev"
      }
   ],
   "keyword":[
      "catalyst"
   ]
}
```


## Right
```json
{
   "node":[
      "Ingredient"
   ],
   "uid":"_:3b45e30f-f15e-4290-8c58-048b846566f9",
   "uuid":"3b45e30f-f15e-4290-8c58-048b846566f9",
   "material":{
      "uuid":"676e3cde-54fa-4dc5-848a-bbd9e229108c"
   },
   "quantity":[
      {
         "node":[
            "Quantity"
         ],
         "uid":"_:b86fb4c3-2a6e-40c0-a0f3-1260e550a51f",
         "uuid":"b86fb4c3-2a6e-40c0-a0f3-1260e550a51f",
         "key":"mass",
         "value":11.2,
         "unit":"kg",
         "uncertainty":0.2,
         "uncertainty_type":"stdev"
      }
   ],
   "keyword":[
      "catalyst"
   ]
}
```

## DB Schema Error
```bash
jsonschema.exceptions.ValidationError: {'node': ['Material'], 'uid': '_:676e3cde-54fa-4dc5-848a-bbd9e229108c', 'uuid': '676e3cde-54fa-4dc5-848a-bbd9e229108c', 'name': 'my test material 0b574777-f154-4e31-9188-8e4f3e516a42_5465890cde6740f1afb4f1b184a2bd07_6814b50b31da4aa9940c44fa3e714415', 'bigsmiles': '123456'} is not valid under any of the given schemas

Failed validating 'anyOf' in schema['properties']['material']:
    {'anyOf': [{'$ref': 'EdgeUUID'}, {'$ref': 'Edge'}]}

On instance['material']:
    {'bigsmiles': '123456',
     'name': 'my test material '
             '0b574777-f154-4e31-9188-8e4f3e516a42_5465890cde6740f1afb4f1b184a2bd07_6814b50b31da4aa9940c44fa3e714415',
     'node': ['Material'],
     'uid': '_:676e3cde-54fa-4dc5-848a-bbd9e229108c',
     'uuid': '676e3cde-54fa-4dc5-848a-bbd9e229108c'}
```

**Comment by nh916** [comment]
I think that was the only issue from our side with the new code and the rest I think are assertion errors that the API is not returning the full JSON.

**Comment by InnocentBug** [comment]
> I think that was the only issue from our side with the new code and the rest I think are assertion errors that the API is not returning the full JSON.

I cannot reproduce this issue.

In fact, when I read in the the "WRONG" json with the SDK.
And print out `cript_nodes.get_json(indent=2).json` I get the "RIGHT" JSON.

**Comment by InnocentBug** [comment]
> > I think that was the only issue from our side with the new code and the rest I think are assertion errors that the API is not returning the full JSON.
> 
> I cannot reproduce this issue.
> 
> In fact, when I read in the the "WRONG" json with the SDK. And print out `cript_nodes.get_json(indent=2).json` I get the "RIGHT" JSON.
> 
> Note that the test function prints out: ` print(my_project_from_api.get_json(sort_keys=False, condense_to_uuid={}, indent=2).json)` which explictly prevents the UUID condensation, which is what makes the WRONG one wrong.

Wait, maybe I see it now.

**Comment by nh916** [comment]
> > I think that was the only issue from our side with the new code and the rest I think are assertion errors that the API is not returning the full JSON.
> 
> I cannot reproduce this issue.
> 
> In fact, when I read in the the "WRONG" json with the SDK. And print out `cript_nodes.get_json(indent=2).json` I get the "RIGHT" JSON.
> 
> Note that the test function prints out: ` print(my_project_from_api.get_json(sort_keys=False, condense_to_uuid={}, indent=2).json)` which explictly prevents the UUID condensation, which is what makes the WRONG one wrong.

okay let me look into it a bit more

**Comment by nh916** [comment]
here is the data that we are sending to the API
```json
{
   "node":[
      "Project"
   ],
   "uid":"_:41904c17-62a2-4b6b-bca2-056d50861c22",
   "uuid":"41904c17-62a2-4b6b-bca2-056d50861c22",
   "name":"test_integration_computation_process_name_ea066ddc580848d0be4cc1a77738d380",
   "collection":[
      {
         "node":[
            "Collection"
         ],
         "uid":"_:cc06092f-0418-450e-ae23-b8db839d7cdc",
         "uuid":"cc06092f-0418-450e-ae23-b8db839d7cdc",
         "name":"my collection name",
         "experiment":[
            {
               "node":[
                  "Experiment"
               ],
               "uid":"_:bcfbc328-ca47-422e-adae-e86249079bbb",
               "uuid":"bcfbc328-ca47-422e-adae-e86249079bbb",
               "name":"my experiment name",
               "computation_process":[
                  {
                     "node":[
                        "ComputationProcess"
                     ],
                     "uid":"_:53769808-6024-4d28-91ad-e499a0c3a8ee",
                     "uuid":"53769808-6024-4d28-91ad-e499a0c3a8ee",
                     "name":"my computational process node name",
                     "type":"cross_linking",
                     "input_data":[
                        {
                           "node":[
                              "Data"
                           ],
                           "uid":"_:4d13f5ec-f0b6-4a1d-99e3-ba0d016754b0",
                           "uuid":"4d13f5ec-f0b6-4a1d-99e3-ba0d016754b0",
                           "name":"my data name",
                           "type":"afm_amp",
                           "file":[
                              {
                                 "node":[
                                    "File"
                                 ],
                                 "uid":"_:0c238567-9b35-40f4-8d2e-40c29f12b461",
                                 "uuid":"0c238567-9b35-40f4-8d2e-40c29f12b461",
                                 "name":"my complex file node fixture",
                                 "source":"https://criptapp.org",
                                 "type":"calibration",
                                 "extension":".csv",
                                 "data_dictionary":"my file's data dictionary"
                              }
                           ]
                        }
                     ],
                     "ingredient":[
                        {
                           "node":[
                              "Ingredient"
                           ],
                           "uid":"_:41b694f9-b4f7-4774-84e7-7253ea808cbd",
                           "uuid":"41b694f9-b4f7-4774-84e7-7253ea808cbd",
                           "material":{
                              "uuid":"0cc8cfe0-a146-48f6-811a-8bab32e9aeef"
                           },
                           "quantity":[
                              {
                                 "node":[
                                    "Quantity"
                                 ],
                                 "uid":"_:8b5118f5-0487-458e-9c67-bc7f837451fa",
                                 "uuid":"8b5118f5-0487-458e-9c67-bc7f837451fa",
                                 "key":"mass",
                                 "value":11.2,
                                 "unit":"kg",
                                 "uncertainty":0.2,
                                 "uncertainty_type":"stdev"
                              }
                           ],
                           "keyword":[
                              "catalyst"
                           ]
                        }
                     ]
                  }
               ],
               "data":[
                  {
                     "uid":"_:4d13f5ec-f0b6-4a1d-99e3-ba0d016754b0"
                  }
               ]
            }
         ],
         "inventory":[
            
         ],
         "citation":[
            
         ]
      }
   ],
   "material":[
      {
         "node":[
            "Material"
         ],
         "uid":"_:0cc8cfe0-a146-48f6-811a-8bab32e9aeef",
         "uuid":"0cc8cfe0-a146-48f6-811a-8bab32e9aeef",
         "name":"my test material b2423383-4173-49f0-a70b-e5b92ac265bf_a8b4c5b2e4954e68ae2416cb194ecab4_8cc43f943b0343bbbff168cd7d63295e",
         "bigsmiles":"123456"
      }
   ]
}
```

is data supposed to be a `UID` or `UUID`?
```json
"data": [
   {
      "uid": "_:4d13f5ec-f0b6-4a1d-99e3-ba0d016754b0"
   }
]
```


```json
"data":[
   {
      "uuid": "4d13f5ec-f0b6-4a1d-99e3-ba0d016754b0"
   }
]
```

**Comment by InnocentBug** [comment]
Nope, that wasn't it.
Now I only have `test_parameter` fail.

**Comment by InnocentBug** [comment]
Let's merge this one now.

The only remaining error is about a patch in `test_parameter`

**Comment by nh916** [comment]
> Let's merge this one now.
> 
> The only remaining error is about a patch in `test_parameter`

but wouldn't this create issues for data?
I think before most of the nodes were working fine, why is this needed?

**Comment by InnocentBug** [comment]
> but wouldn't this create issues for data?
> I think before most of the nodes were working fine, why is this needed?

Before we were not compliant with the idea of UIDs.
Although the data schema did not complain, we overspecified the `node` type.

A consequence was that we couldn't read in JSON generated outside the SDK. (Berengers scripts.)

So, this is necessary.

**Comment by nh916** [comment]
> > but wouldn't this create issues for data?
> > I think before most of the nodes were working fine, why is this needed?
> 
> Before we were not compliant with the idea of UIDs. Although the data schema did not complain, we overspecified the `node` type.
> 
> A consequence was that we couldn't read in JSON generated outside the SDK. (Berengers scripts.)
> 
> So, this is necessary.

I'm not sure what to do because this is needed to read in JSON into the SDK, but if we add this right before ACS, we'd be introducing bugs into our SDK.

What are your thoughts on merging this in after ACS?

**Comment by nh916** [comment]
tested this locally and it seemed to pass with staging

**Comment by InnocentBug** [comment]
> since it is passing I am more confident to merge this in, but if it is not fixing any immediate bugs I am also happy to leave it and merge it next week after ACS

it does fix an immediate bug!

You can't read in JSON with UID from back end or Berenger right now.

You need this fix for it.

**Review by nh916 — APPROVED** [review]
